packages/components/src/validated-form-controls/control-with-error.tsx

```diff
@@ -0,0 +1,161 @@
+import { Icon } from '@wordpress/components';
+import { __ } from '@wordpress/i18n';
+import { error } from '@wordpress/icons';
+import { cloneElement, forwardRef, useEffect, useState } from 'react';
+import { withIgnoreIMEEvents } from '../utils/with-ignore-ime-events';
+
+import './style.scss';
+
+function appendRequiredIndicator(
+	label: string,
+	required: boolean | undefined,
+	markWhenOptional: boolean | undefined
+) {
+	if ( required && ! markWhenOptional ) {
+		return `${ label } (${ __( 'Required' ) })`;
+	}
+	if ( ! required && markWhenOptional ) {
+		return `${ label } (${ __( 'Optional' ) })`;
+	}
+	return label;
+}
+
+/**
+ * HTML elements that support the Constraint Validation API.
+ *
+ * Here, we exclude HTMLButtonElement because although it does technically support the API,
+ * normal buttons are actually exempted from any validation.
+ * @see https://developer.mozilla.org/en-US/docs/Learn_web_development/Extensions/Forms/Form_validation
+ * @see https://developer.mozilla.org/en-US/docs/Web/API/HTMLButtonElement/willValidate
+ */
+type ValidityTarget =
+	| HTMLFieldSetElement
+	| HTMLInputElement
+	| HTMLSelectElement
+	| HTMLTextAreaElement;
+
+function UnforwardedControlWithError< C extends React.ReactElement >(
+	{
+		required,
+		markWhenOptional,
+		customValidator,
+		getValidityTarget,
+		render,
+	}: {
+		/**
+		 * Whether the control is required.
+		 */
+		required?: boolean;
+		/**
+		 * Label the control as "optional" when _not_ `required`, instead of the inverse.
+		 */
+		markWhenOptional?: boolean;
+		/**
+		 * A function that returns a custom validity message when applicable.
+		 *
+		 * This message will be applied to the element returned by `getValidityTarget`.
+		 * @see https://developer.mozilla.org/en-US/docs/Web/API/HTMLObjectElement/setCustomValidity
+		 */
+		customValidator?: () => string | void;
+		/**
+		 * A function that returns the actual element on which the validity data should be applied.
+		 */
+		getValidityTarget: () => ValidityTarget | null | undefined;
+		render: C;
+	},
+	forwardedRef: React.ForwardedRef< HTMLDivElement >
+) {
+	const [ errorMessage, setErrorMessage ] = useState< string | undefined >();
+	const [ isTouched, setIsTouched ] = useState( false );
+
+	// Ensure that error messages are visible after user attemps to submit a form
+	// with multiple invalid fields.
+	useEffect( () => {
+		const validityTarget = getValidityTarget();
+		const showValidationMessage = () => setErrorMessage( validityTarget?.validationMessage );
+
+		validityTarget?.addEventListener( 'invalid', showValidationMessage );
+
+		return () => {
+			validityTarget?.removeEventListener( 'invalid', showValidationMessage );
+		};
+	} );
```

Comment:
Should we add dependencies for useEffect?

Reply:
This effect wants to capture the validation state on each state change of the control component, so I feel safer to run it on all updates (as long as there are no perceivable performance issues 😄).